geo: remove pointer usage for Geometry/Geography

otan opened the pull request:
This should improve our heap profiles by making tree.DGeometry/Geography
only implant a geo.Geometry/Geography object.

Release justification: low risk, high reward changes to existing
functionality

Release note: None

otan committed:
geo: remove pointer usage for Geometry/Geography

This should improve our heap profiles by making tree.DGeometry/Geography
only implant a geo.Geometry/Geography object.

Release justification: low risk, high reward changes to existing
functionality

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/53446)
<!-- Reviewable:end -->

nvanbenschoten commented on pkg/geo/geogfn/dwithin.go:
```diff
@@ -21,14 +21,14 @@ import (
 ) (bool, error) {
 	if a.SRID() != b.SRID() {
-		return false, geo.NewMismatchingSRIDsError(a, b)
+		return false, geo.NewMismatchingSRIDsError(&a, &b)
```
Drive-by comment: this error is probably making the local vars escape back to the heap wherever it’s used. We might want to make the args values too.

otan replied:
Might address that in a follow up PR. Gotta remove the interface and use spatial object instead.

nvanbenschoten replied:
SGTM. I did just confirm that this is the case though, so you'll want to make that change before running `sqlcmp`. As is, this might actually be a pessimization.

nvb reviewed (commented):
Nice!

sumeerbhola reviewed (approved):
Looks good.

- The `sizeof` a `SpatialObject` is currently 48 bytes. With some reduction in size of fields and rearrangement we can bring it down to 40 bytes.
SRID: int32
ShapeType: int32 but can be changed to int16
Type: ditto
EWKB: 24 bytes
BoundingBox: 8 bytes
If we reorder the fields as above, the first 3 fields get packed into an 8 byte word (the old ordering adds padding so no improvement).

- Can you run `sqlcmp` once with these changes just to make sure we don't have an unforeseen regression.

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 
 

> Reviewed 57 of 57 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53446#-:-MFfZiJwESzlNHaBfoNC:b55lczz)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @nvanbenschoten)



<!-- Sent from Reviewable.io -->

otan commented:
> The sizeof a SpatialObject is currently 48 bytes. With some reduction in size of fields and rearrangement we can bring it down to 40 bytes.

Does that change the encoding, or is protobuf okay with downgrading int sizes?

nvanbenschoten commented:
> EWKB: 24 bytes

If you're always parsing this using `ewkb.Read`, you can shave off 8 bytes by switching this to a `string` (e.g. an immutable byte slice). For this to work, you'd need to use a [`strings.Reader`](https://golang.org/pkg/strings/#Reader) instead of the `bytes.Reader` we talked about before.

Doing so it likely not worth it if means you'd need to go from `[]byte` to `string` and back often.

otan commented:
good call @nvanbenschoten, after fixing the mismatching SRID stuff:

```
name                                   old ms     new ms     delta
Test/postgis_geometry_tutorial/11.2b   3.88 ± 6%  2.88 ± 9%  -25.87%  (p=0.008 n=5+5)
Test/postgis_geometry_tutorial/11.6    3.80 ±11%  3.27 ±18%     ~     (p=0.222 n=5+5)
Test/postgis_geometry_tutorial/12.1.2  3.67 ±16%  2.15 ±16%  -41.34%  (p=0.008 n=5+5)
Test/postgis_geometry_tutorial/12.2.3  3.77 ±15%  2.98 ±21%  -21.11%  (p=0.032 n=5+5)
Test/postgis_geometry_tutorial/12.2.4  3.48 ±11%  3.44 ±67%     ~     (p=0.310 n=5+5)
Test/postgis_geometry_tutorial/13.0    4.05 ± 4%  3.17 ±18%  -21.67%  (p=0.008 n=5+5)
Test/postgis_geometry_tutorial/13.1a    428 ± 2%   412 ±13%     ~     (p=0.151 n=5+5)
Test/postgis_geometry_tutorial/13.1c   46.5 ± 8%  47.4 ±13%     ~     (p=0.841 n=5+5)
Test/postgis_geometry_tutorial/13.2     540 ± 4%   529 ± 8%     ~     (p=0.310 n=5+5)
Test/postgis_geometry_tutorial/14.1a   6.48 ± 5%  3.58 ± 7%  -44.77%  (p=0.008 n=5+5)
Test/postgis_geometry_tutorial/14.2b   19.8 ±35%  12.0 ±14%  -39.57%  (p=0.008 n=5+5)
Test/postgis_geometry_tutorial/14.2c   10.7 ±13%   6.8 ± 1%  -36.70%  (p=0.016 n=5+4)
Test/postgis_geometry_tutorial/14.3c   55.8 ±19%  47.7 ± 9%  -14.60%  (p=0.032 n=5+5)
Test/postgis_geometry_tutorial/15.0    4.11 ±22%  3.86 ±28%     ~     (p=0.421 n=5+5)
```

tacked on as a separate commit.

otan committed:
geo: use SpatialObject for MismatchingSRIDError

Release justification: low risk, high benefit changes to existing functionality
Release note: None

sumeerbhola reviewed (commented):
> Does that change the encoding, or is protobuf okay with downgrading int sizes?

they are all varint encoded so should be ok, and we don't care about compatibility yet. But I overlooked two things: these are enums in the proto file; because of the varint encoding, there is no int16 type for protos. Maybe a proto extension can be used to tell gogoproto to do the right thing -- like `gogoproto.casttype`? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/53446#-:-MFg-tje4MmI4lRJSkXr:b-s9d8s2)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @otan and @sumeerbhola)



<!-- Sent from Reviewable.io -->

otan commented:
bors r=sumeerbhola

> they are all varint encoded so should be ok, and we don't care about compatibility yet

I'm not sure -- I think we care once beta hits.

> Maybe a proto extension can be used to tell gogoproto to do the right thing -- like gogoproto.casttype?

I think we can do that.

> BoundingBox: 8 bytes
This escapes to the heap right now. If we can "flatten" it that may improve performance, but I'd need to be able "NULL" bounding box (maybe a bool or something).

otan commented:
bors r=sumeerbhola

otan commented:
bors r=sumeerbhola

sumeerbhola commented:
> The sizeof a SpatialObject is currently 48 bytes. With some reduction in size of fields and rearrangement we can bring it down to 40 bytes.

btw, I tried this out but it made no difference to query latency
https://github.com/sumeerbhola/cockroach/commit/b688736e328d02094d7b479ac16ac8fe0843a6ff
